### PR TITLE
[Model] Add Qwen2 model.

### PR DESCRIPTION
**Pull request by marvin-Yu**

### CODE REVIEW

**Comment by changqi1** [comment]
What is the differences between Qwen1.0 vs Qwen 2.0?

File: src/models/qwen2.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+private:
+    void setEmbeddingWeights(const std::string &modelPath);
+    void setFinalLnWeight(const std::string &modelPath);
+
+private:
+    TokenEmbedding<float16_t> *embedding;
+    RmsNorm finalLN;
+};
+
+REGISTER_DECODER(Qwen2LLM, qwen2, float)
```
Comment (pujiang2018) [inline src/models/qwen2.h:52]: What's the error when put such macro to cpp file?

Comment (Duyi-Wang) [reply src/models/qwen2.h:52]: > What's the error when put such macro to cpp file?

Putting in a CPP file won't trigger the execution of this macro. It instantiates a static object of a registrar.

File: benchmark/run.sh [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-# set -x
```
Comment (Duyi-Wang) [inline benchmark/run.sh:2]: unset

File: examples/model_config/qwen2-14b/model.safetensors.index.json
Comment (Duyi-Wang) [inline examples/model_config/qwen2-14b/model.safetensors.index.json:1]: unnecessary

File: examples/model_config/qwen2-4b/model.safetensors.index.json
Comment (Duyi-Wang) [inline examples/model_config/qwen2-4b/model.safetensors.index.json:1]: Unnecessary

File: examples/model_config/qwen2-72b/model.safetensors.index.json
Comment (Duyi-Wang) [inline examples/model_config/qwen2-72b/model.safetensors.index.json:1]: Unnecessary

File: examples/model_config/qwen2-7b/model.safetensors.index.json
Comment (Duyi-Wang) [inline examples/model_config/qwen2-7b/model.safetensors.index.json:1]: Unnecessary

File: examples/model_config/qwen2-0_5b/config.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  ],
+  "attention_dropout": 0.0,
+  "bos_token_id": 151643,
+  "eos_token_id": 151645,
+  "hidden_act": "silu",
+  "hidden_size": 1024,
+  "initializer_range": 0.02,
+  "intermediate_size": 2816,
+  "max_position_embeddings": 32768,
+  "max_window_layers": 21,
```
Comment (changqi1) [inline examples/model_config/qwen2-0_5b/config.json:13]: what is the meaning for max_window_layers?

**Review by Duyi-Wang — COMMENTED** [review]
copy right's year

**Comment by changqi1** [comment]
> What is the differences between Qwen1.0 vs Qwen 2.0?

If Qwen2.0 have no difference w/ LLama2. Do we need to inherit from LLama2?

File: src/models/qwen2.h (hunk 1 above)
Comment (marvin-Yu) [reply src/models/qwen2.h:52]: not the scope of this PR.

File: examples/model_config/qwen2-0_5b/config.json (hunk 3 above)
Comment (marvin-Yu) [reply examples/model_config/qwen2-0_5b/config.json:13]: Done

File: examples/model_config/qwen2-4b/model.safetensors.index.json
Comment (marvin-Yu) [reply examples/model_config/qwen2-4b/model.safetensors.index.json:1]: Done

File: examples/model_config/qwen2-72b/model.safetensors.index.json
Comment (marvin-Yu) [reply examples/model_config/qwen2-72b/model.safetensors.index.json:1]: Done

File: examples/model_config/qwen2-7b/model.safetensors.index.json
Comment (marvin-Yu) [reply examples/model_config/qwen2-7b/model.safetensors.index.json:1]: Done

**Review by changqi1 — APPROVED** [review]

**Pull request merged**